packages/phpstan-rules/src/Rules/RequireStringRegexMatchKeyRule.php

```diff
@@ -0,0 +1,184 @@
+use Symplify\PHPStanRules\ValueObject\PHPStanAttributeKey;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @see \Symplify\PHPStanRules\Tests\Rules\RequireStringRegexMatchKeyRule\RequireStringRegexMatchKeyRuleTest
+ */
+final class RequireStringRegexMatchKeyRule extends AbstractSymplifyRule
+{
+    /**
+     * @var string
+     */
+    public const ERROR_MESSAGE = '"%s" regex need to use string named capture group instead of numeric';
+
+    /**
+     * @var NodeFinder
+     */
+    private $nodeFinder;
+
+    /**
+     * @var NodeComparator
+     */
+    private $nodeComparator;
+
+    public function __construct(NodeFinder $nodeFinder, NodeComparator $nodeComparator)
+    {
+        $this->nodeFinder = $nodeFinder;
+        $this->nodeComparator = $nodeComparator;
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        return [ArrayDimFetch::class];
+    }
+
+    /**
+     * @param ArrayDimFetch $node
+     * @return string[]
+     */
+    public function process(Node $node, Scope $scope): array
+    {
+        $dim = $node->dim;
+        if (! $dim instanceof LNumber) {
+            return [];
+        }
+
+        $regexMatchAssign = $this->getRegexMatchAssign($node);
+        if ($regexMatchAssign === null) {
+            return [];
+        }
+
+        /** @var StaticCall $expr */
+        $expr = $regexMatchAssign->expr;
+        /** @var ClassConstFetch $value */
+        $value = $expr->args[1]->value;
+        $regex = (string) $value->getAttribute(PHPStanAttributeKey::PHPSTAN_CACHE_PRINTER);
+
+        return [sprintf(self::ERROR_MESSAGE, $regex)];
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(self::ERROR_MESSAGE, [
+            new CodeSample(
+                <<<'CODE_SAMPLE'
+use Nette\Utils\Strings;
+
+class SomeClass
+{
+    private const REGEX = '#(a content)#';
+
+    public function run()
+    {
+        $matches = Strings::match('a content', self::REGEX);
+        if ($matches) {
+            echo $matches[1];
+        }
+    }
+}
+CODE_SAMPLE
+                ,
+                <<<'CODE_SAMPLE'
+use Nette\Utils\Strings;
+
+class SomeClass
+{
+    private const REGEX = '#(?<c>a content)#';
+
+    public function run()
+    {
+        $matches = Strings::match('a content', self::REGEX);
+        if ($matches) {
+            echo $matches['c'];
+        }
+    }
+}
+CODE_SAMPLE
+            ),
+        ]);
+    }
+
+    private function getRegexMatchAssign(ArrayDimFetch $arrayDimFetch): ?Assign
+    {
+        $parent = $arrayDimFetch->getAttribute(PHPStanAttributeKey::PARENT);
+        while ($parent) {
+            $previous = $parent->getAttribute(PHPStanAttributeKey::PREVIOUS);
+            while ($previous) {
+                $assign = $this->getArrayDimFetchAssign($previous, $arrayDimFetch);
```

Comment:
While loops are potential segfauls and make code readable.

The parent method should be used here instead: https://github.com/symplify/symplify/blob/master/packages/phpstan-rules/src/Rules/AbstractSymplifyRule.php#L56

Reply:
The use case is different, it want to locate the nearest assign on previous statement as variable can be re-assign before, also not looking assign after usage, if not found, look parent again and check previous.

Reply:
I understand that. It's maybe rather quesiton of best practise in traversing.

It's better to get a node, then traverse it down. Traversing up can go wrong and is inconsistent. E.g. in Rector, node traversed up cannot be always changed due to php-parser design, so if you try to change it, it can cause bugs.


In this case it would be better to listen to `Assign` node, then look bellow under it for it's left variable array dim fetches, e.g.:


```php
// Assign
$matches = Strings::match('...');


if ($matches[0] ...) {
}
```

Reply:
I will try.

Reply:
implemented.

Reply:
Thank you